Feature optimizations and refactorings for anime route

delikesance opened the pull request:
This merge includes a series of significant changes aimed at optimizing the `/animes` route, introducing caching mechanisms, enhancing query handling, and refining the API's structure for better performance and maintainability. Major features and refactorings are highlighted below:

- **feat(animes-route):** Optimized anime list fetching with caching and improved query handling, introducing a cache-first strategy to enhance performance significantly. Adjusted query parameter handling for filtering animes by title, genres, status, and release date to ensure more efficient database queries. Also included streamlined handling for fetching detailed anime information when necessary and adjusted the main application logger settings for development purposes. Note that these changes might affect existing queries due to the adjustments in query parameter handling, marking a breaking change.

- **refactor:** Various refactorings across the codebase to improve code quality and organization. This includes marking plugin and route default exports as async, improving query construction with better type safety, and removing redundant comments in favor of relying on TypeScript's typing system and external API documentation.

- **feat:** Introduced and enhanced several services (`AnimeService`, `UserService`) for centralized data handling and operations related to anime data and user registration/authentication, respectively.

- **BREAKING CHANGE:** Transitioned from using Prettier and ESLint to Biome for code formatting and linting to align with new development standards.

These changes are expected to improve the scalability, performance, and maintainability of the gazes-api, setting a solid foundation for future enhancements.

PR by Angelo Clauin <contact@l31.dev>

delikesance committed:
refactor: Remove redundant route comments

Removed detailed JSDoc comments from the animes.ts route handlers. This
cleanup aims to rely more on Typescript's typing system and external API
documentation for clarity.

delikesance committed:
refactor: Improve query construction

- Introduced Prisma type import for better type safety in database query
  construction
- Reorganized imports for better readability and maintenance, ensuring a
  logical order and grouping
- Refactored the anime listing query options to be constructed more
  directly, enhancing code clarity and reducing complexity
- Annotated 'queryOptions' with a specific Prisma type to enforce
  structure and type safety

delikesance committed:
feat(animes.ts): Enhance anime list query

- Modified the anime listing endpoint to include related episodes in the
  database query results by adding an 'include' option to the Prisma query.

- Implemented additional logic to enrich anime entries lacking a synopsis with
  detailed information by fetching it via `animeService.getAnimeById`. This
  ensures the response provides a more comprehensive dataset, improving the
  frontend user experience by providing episode details and ensuring all anime
  listings have synopses.

delikesance committed:
refactor: Mark plugin and route default exports as async

Updated the default export functions in the `sensible.ts` plugin and
`auth.ts`, `root.ts` routes to be asynchronous. This change allows for
the use of async-await syntax within these functions, preparing the
ground for potential asynchronous operations like database queries or
external API calls. Although the current implementation does not
immediately utilize the async capabilities, this adjustment ensures that
the API is ready for future asynchronous enhancements without further
structural changes.

delikesance committed:
feat(animes-route): optimize anime list fetching with caching and
improve query handling

Introduce caching for the anime list route to improve performance by
attempting to fetch anime lists from cache before querying the database.
This update optimizes the handling of query parameters for filtering
animes by title, genres, status, and release date, ensuring more
efficient database queries. Additionally, refine the handling of missing
anime synopses by fetching detailed information in a more streamlined
manner. Also, adjust the main application logger settings for
development purposes.

BREAKING CHANGE: Adjustments in query parameter handling could affect
existing queries relying on previous behavior.

ncs-pl reviewed (commented):
lgtm, appart from the already existing issues from automatic static analysis.
I would also encourage you avoiding "magic values", and instead declare constants that are being inlined at compile time, for illustration:

```ts
response.status(200)...;
```

to be changed to

```ts
const HTTP_OK = 200;

response.status(HTTP_OK)...;
```

lilmaax commented on apps/gazes-api/src/app/routes/animes.ts:
```diff
@@ -1,56 +1,57 @@
-				equals: releaseDate.toString(),
-			};
+		let animeList = await prisma.anime.findMany(queryOptions);
+		await redis.setEx(cacheKey, 3600, JSON.stringify(animeList));
```
What does "3600" value mean? Maybe you can set an explicit variable instead

maxchrr reviewed (commented):
lgtm, Code Quality analysis has already done all the job

delikesance committed:
feat: implement required environment variable validation

This update introduces a robust mechanism for validating the presence of
required environment variables within the application's configuration
setup. The `getEnv` function now throws an error if a necessary
environment variable is not defined, enhancing the app's stability by
ensuring all critical configurations are loaded at startup.

This change solidifies the configuration loading mechanism, making it
more reliable and easier to debug in environments where configurations
might be misconfigured or omitted.

delikesance committed:
feat(server-setup): integrate Fastify, Prisma, and Redis

Enhanced the application's architecture by integrating Fastify for the
web server framework, Prisma as the ORM for database interactions, and
Redis for caching capabilities. This setup provides a robust foundation
for building scalable web applications.

- Initialized Fastify with custom configuration, disabling the logger by
  default for streamlined operation.
- Established a PrismaClient instance for database interactions,
  enabling advanced data management and query capabilities.
- Configured Redis client and implemented automatic connection logic,
  including environment-specific behavior (e.g., flushing the database
  in development mode for a clean slate).
- Defined `AppOptions` interface to type-check application options,
  ensuring type safety across app components.
- Set up graceful shutdown procedures to properly close database and
  Redis connections upon application termination, enhancing resource
  management and stability.
- Refined the server's listen method to dynamically use environment
  variables for port and host settings, improving deployment
  flexibility.

This commit lays the groundwork for a modular, efficient backend
structure that leverages the strengths of each integrated technology.

delikesance committed:
Refactor(animeService): Streamline anime data operations and caching

This commit significantly refactors the animeService class to enhance
readability, improve cache management, and remove unused imports and
functions. Key changes include:

- Removed unused imports such as `config`, `PrismaClient`, and various
  types from "@api/contracts/animesContract".
- Simplified the `AnimeService` constructor to use `PrismaClient` and
  `CacheManager` directly, enhancing dependency injection and testing
  capabilities.
- Optimized `getAnimeById` to better utilize caching for anime details,
  reducing unnecessary database queries.
- Introduced `getAndEnrichAnimesList` to fetch and optionally enrich a
  list of animes with additional details, leveraging both database
  queries and cache for improved performance.
- Removed outdated functions responsible for updating the anime database
  and fetching the latest episodes. This functionality is presumed to be
  handled elsewhere or refactored into a more appropriate service or
  scheduled task.
- Updated routes in `animes.ts` to reflect changes in the animeService's
  API, ensuring compatibility with the refactored service methods.
- Adjusted other service classes and route handlers to align with the
  refactoring, ensuring consistent use of `PrismaClient` and
  `CacheManager`.

Overall, this commit aims to make the animeService more efficient,
maintainable, and aligned with best practices in cache usage and data
fetching.

garder500 committed:
Adding two new modules (and removing a deprecated module openssl)

garder500 committed:
Typing Schema changed to use "TypeBox" (More efficient, and less work)

garder500 committed:
Removing the JwsUtils.ts

garder500 committed:
Add of the plugins jwt.ts with specifieds handlers

garder500 committed:
Changing minors thing their (adding atest caching and puting schema to use Timestamp in database)

garder500 committed:
Add a route to logged in Users (Using a Bearer Token of course)

lilmaax commented on apps/gazes-api/src/app/plugins/jwt.ts:
```diff
@@ -0,0 +1,16 @@
+ *
+ * @see https://github.com/fastify/fastify-sensible
+ */
+export default fp(async (fastify: FastifyInstance) => {
+	fastify.register(jwt, {
+        secret: getEnv("JWT_SECRET")
+    });
+    
+});
```
Indentation seems weird

lilmaax commented on apps/gazes-api/src/app/routes/animes.ts:
```diff
@@ -1,93 +1,49 @@
+import { AnimeDetailParams, AnimeDetailParamsSchema, AnimeListQuerystring, AnimeListQuerystringSchema } from "@api/contracts/animesContract";
+import { StatusCodes } from "http-status-codes";
+import { Anime, Prisma } from "@prisma/client";
+import { CacheManager } from "../utils/cacheManager";
```
Try not to use relative import, if you move the file or implementation, you can break it

lilmaax commented on apps/gazes-api/src/app/routes/animes.ts:
```diff
@@ -98,48 +54,6 @@ export default async function (fastify: FastifyInstance, { redis, prisma }: AppO
 
 		rep.status(200).send(anime);
```
Missing `StatusCodes` variable

lilmaax commented on apps/gazes-api/src/app/routes/auth.ts:
```diff
@@ -24,7 +24,7 @@ export default function (fastify: FastifyInstance, { prisma }: AppOptions) {
+			const { user, token } = await userService.registerUser(email, username, password, fastify);
 
 			rep.header("Set-Cookie", `token=${token}; HttpOnly; Path=/; Secure; SameSite=Strict`);
 			rep.status(201).send(user);
```
Missing `StatusCodes` variable

lilmaax commented on apps/gazes-api/src/app/routes/auth.ts:
```diff
@@ -50,7 +50,7 @@ export default function (fastify: FastifyInstance, { prisma }: AppOptions) {
+			const { user, token } = await userService.authenticateUser(email, password, fastify);
 
 			rep.header("Set-Cookie", `token=${token}; HttpOnly; Path=/; Secure; SameSite=Strict`);
 			rep.status(200).send(user);
```
Missing `StatusCodes` variable

lilmaax commented on apps/gazes-api/src/app/services/animeService.ts:
```diff
@@ -1,123 +1,109 @@
+		releaseDate,
+	}: { page: number; title: string; genres: string; status: number; releaseDate: number }): Prisma.AnimeFindManyArgs {
+		return {
+			skip: 25 * (page - 1),
+			take: 25,
```
Maybe you can make this more readable, or explain it with comment

lilmaax commented on apps/gazes-api/src/main.ts:
```diff
@@ -1,46 +1,94 @@
-		redis.flushDb();
-	}
+	updateAnimeDatabase(prismaClient);
+	setInterval(() => updateAnimeDatabase(prismaClient), 3600000);
```
Use constant instead of set value, and explain what it means

maxchrr reviewed (commented):
More generally, there seem to be indentation/formatting issues in a few lines, which is very weird.

garder500 committed:
Fix WHAT MAX SAID

garder500 committed:
Finalisation of The API (need a Final review, but everything should work)

maxchrr reviewed (commented):
LGTM

Don't forget to add comments, everything is very rough, to help you understand the code.

There also seem to be some issues with the rule for spaces between braces, so try to be consistent.
